src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -58,35 +86,123 @@ internal sealed class ExternalErrorDiagnosticUpdateSource : IDiagnosticUpdateSou
+            // on a serialized task queue.
+            _taskQueue.ScheduleTask(nameof(ClearErrors), async () =>
             {
-                // this will get called if the project is actually built by "build" command.
-                // we track what project has been built, so that later we can clear any stale live errors
-                // if build give us no errors for this project
-                state?.Built(projectId);
+                if (state == null)
+                {
+                    // TODO: Is it possible that ClearErrors can be invoked while the build is not in progress?
+                    // We fallback to current solution in the workspace and clear errors for the project.
+                    await ClearErrorsCoreAsync(projectId, _workspace.CurrentSolution, state).ConfigureAwait(false);
+                }
+                else
+                {
+                    // We are going to clear the diagnostics for the current project.
+                    // Additionally, we clear errors for all projects that transitively depend on this project.
+                    // Otherwise, fixing errors in core projects in dependency chain will leave back stale diagnostics in dependent projects.
 
-                ClearProjectErrors(state?.Solution ?? _workspace.CurrentSolution, projectId);
+                    // First check if we already cleared the diagnostics for this project when processing a referenced project.
+                    // If so, we don't need to clear diagnostics for it again.
+                    if (state.WereProjectErrorsCleared(projectId))
+                    {
+                        return;
+                    }
+
+                    var solution = state.Solution;
+
+                    await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
+
+                    var transitiveProjectIds = solution.GetProjectDependencyGraph().GetProjectsThatTransitivelyDependOnThisProject(projectId);
+                    foreach (var projectId in transitiveProjectIds)
+                    {
+                        if (state.WereProjectErrorsCleared(projectId))
+                        {
+                            continue;
+                        }
+
+                        await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
+                    }
+                }
             }, CancellationToken.None);
+
+            return;
+
+            async Task ClearErrorsCoreAsync(ProjectId projectId, Solution solution, InProgressState? state)
+            {
+                Debug.Assert(state == null || !state.WereProjectErrorsCleared(projectId));
+
+                // Here, we clear the build and live errors for the project.
+                // Additionally, we mark projects as having its errors cleared.
+                // This ensures that we do not attempt to clear the diagnostics again for the same project
```

Comment:
"This ensures that we do not attempt to clear the diagnostics again for the same project"

i'm curious why that's important for us to do.  if we clear twice, what's the issue?

Reply:
Just a performance overhead, we will do unnecessary compute, raise diagnostic removed events and then the build + intellisense diagnostic sources will attempt to clear diagnostics, when we know none exist as we already cleared them. Let me add a PERF prefix to the comment.

Reply:
Actually, given the only change needed for the new feedback is an improved comment, I will do it in a follow-up PR to avoid re-triggering all tests. I plan to continue iterating on improving this space.